src/app/tests/TestReadInteraction.cpp

```diff
@@ -0,0 +1,190 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements unit tests for CHIP Interaction Model Read Interaction
+ *
+ */
+
+#include <app/InteractionModelEngine.h>
+#include <core/CHIPCore.h>
+#include <core/CHIPTLV.h>
+#include <core/CHIPTLVDebug.hpp>
+#include <core/CHIPTLVUtilities.hpp>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <messaging/Flags.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <support/ErrorStr.h>
+#include <support/UnitTestRegistration.h>
+#include <system/SystemPacketBuffer.h>
+#include <system/TLVPacketBufferBackingStore.h>
+#include <transport/PASESession.h>
+#include <transport/SecureSessionMgr.h>
+#include <transport/raw/UDP.h>
+
+#include <nlunit-test.h>
+
+namespace chip {
+SecureSessionMgr gSessionManager;
+Messaging::ExchangeManager gExchangeManager;
+TransportMgr<Transport::UDP> gTransportManager;
+const Transport::AdminId gAdminId = 0;
+
+namespace app {
+class TestReadInteraction
+{
+public:
+    static void TestReadClient(nlTestSuite * apSuite, void * apContext);
+    static void TestReadHandler(nlTestSuite * apSuite, void * apContext);
+
+private:
+    static void GenerateReportData(nlTestSuite * apSuite, void * apContext, System::PacketBufferHandle aPayload);
+};
+
+void TestReadInteraction::GenerateReportData(nlTestSuite * apSuite, void * apContext, System::PacketBufferHandle aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVWriter writer;
+    writer.Init(std::move(aPayload));
+
+    ReportData::Builder reportDataBuilder;
+
+    err = reportDataBuilder.Init(&writer);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    reportDataBuilder.SuppressResponse(true);
+    NL_TEST_ASSERT(apSuite, reportDataBuilder.GetError() == CHIP_NO_ERROR);
+
+    reportDataBuilder.MoreChunkedMessages(false);
+    NL_TEST_ASSERT(apSuite, reportDataBuilder.GetError() == CHIP_NO_ERROR);
+
+    reportDataBuilder.EndOfReportData();
+    NL_TEST_ASSERT(apSuite, reportDataBuilder.GetError() == CHIP_NO_ERROR);
+
+    err = writer.Finalize(&aPayload);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+}
+
+void TestReadInteraction::TestReadClient(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    app::ReadClient readClient;
+
+    System::PacketBufferHandle buf = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    err                            = readClient.Init(&gExchangeManager, nullptr);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    err = readClient.SendReadRequest(kTestDeviceNodeId, gAdminId);
+    NL_TEST_ASSERT(apSuite, err == CHIP_ERROR_INCORRECT_STATE);
+
+    GenerateReportData(apSuite, apContext, buf.Retain());
+
+    err = readClient.ProcessReportData(std::move(buf));
```

Comment:
This still seems like a weird thing to do.  We're bypassing some important parts of the normal message-processing codepath by doing things this way.  I don't want to block on this, so followup OK, but maybe we should have a public API that returns error that does what OnMessageReceived does right now.  OnMessageReceived would call it, and we could call it from here....

Reply:
thanks, I would follow up it in https://github.com/project-chip/connectedhomeip/issues/5411